// ra-data-json-server/dist/esm/index.js
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (g && (g = 0, op[0] && (_ = 0)), _) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
import { stringify } from 'query-string';
import { fetchUtils } from 'ra-core';
/**
 * Maps react-admin queries to a json-server powered REST API
 *
 * @see https://github.com/typicode/json-server
 *
 * @example
 *
 * getList          => GET http://my.api.url/posts?_sort=title&_order=ASC&_start=0&_end=24
 * getOne           => GET http://my.api.url/posts/123
 * getManyReference => GET http://my.api.url/posts?author_id=345
 * getMany          => GET http://my.api.url/posts?id=123&id=456&id=789
 * create           => POST http://my.api.url/posts/123
 * update           => PUT http://my.api.url/posts/123
 * updateMany       => PUT http://my.api.url/posts/123, PUT http://my.api.url/posts/456, PUT http://my.api.url/posts/789
 * delete           => DELETE http://my.api.url/posts/123
 *
 * @example
 *
 * import * as React from "react";
 * import { Admin, Resource } from 'react-admin';
 * import jsonServerProvider from 'ra-data-json-server';
 *
 * import { PostList } from './posts';
 *
 * const App = () => (
 *     <Admin dataProvider={jsonServerProvider('http://jsonplaceholder.typicode.com')}>
 *         <Resource name="posts" list={PostList} />
 *     </Admin>
 * );
 *
 * export default App;
 */
export default (function (apiUrl, httpClient) {
    if (httpClient === void 0) { httpClient = fetchUtils.fetchJson; }
    return ({
        getList: function (resource, params) { return __awaiter(void 0, void 0, void 0, function () {
            var _a, page, perPage, _b, field, order, query, url, _c, headers, json, totalString;
            var _d;
            return __generator(this, function (_e) {
                switch (_e.label) {
                    case 0:
                        _a = params.pagination || {}, page = _a.page, perPage = _a.perPage;
                        _b = params.sort || {}, field = _b.field, order = _b.order;
                        query = __assign(__assign({}, fetchUtils.flattenObject(params.filter)), { _sort: field, _order: order, _start: page != null && perPage != null
                                ? (page - 1) * perPage
                                : undefined, _end: page != null && perPage != null ? page * perPage : undefined, _embed: (_d = params === null || params === void 0 ? void 0 : params.meta) === null || _d === void 0 ? void 0 : _d.embed });
                        url = "".concat(apiUrl, "/").concat(resource, "?").concat(stringify(query));
                        return [4 /*yield*/, httpClient(url, {
                                signal: params === null || params === void 0 ? void 0 : params.signal,
                            })];
                    case 1:
                        _c = _e.sent(), headers = _c.headers, json = _c.json;
                        if (!headers.has('x-total-count')) {
                            throw new Error('The X-Total-Count header is missing in the HTTP Response. The jsonServer Data Provider expects responses for lists of resources to contain this header with the total number of results to build the pagination. If you are using CORS, did you declare X-Total-Count in the Access-Control-Expose-Headers header?');
                        }
                        totalString = headers.get('x-total-count').split('/').pop();
                        if (totalString == null) {
                            throw new Error('The X-Total-Count header is invalid in the HTTP Response.');
                        }
                        return [2 /*return*/, { data: json, total: parseInt(totalString, 10) }];
                }
            });
        }); },
        getOne: function (resource, params) { return __awaiter(void 0, void 0, void 0, function () {
            var url, json;
            var _a;
            return __generator(this, function (_b) {
                switch (_b.label) {
                    case 0:
                        url = "".concat(apiUrl, "/").concat(resource, "/").concat(params.id);
                        if ((_a = params === null || params === void 0 ? void 0 : params.meta) === null || _a === void 0 ? void 0 : _a.embed) {
                            url += "?_embed=".concat(params.meta.embed);
                        }
                        return [4 /*yield*/, httpClient(url, { signal: params === null || params === void 0 ? void 0 : params.signal })];
                    case 1:
                        json = (_b.sent()).json;
                        return [2 /*return*/, { data: json }];
                }
            });
        }); },
        getMany: function (resource, params) { return __awaiter(void 0, void 0, void 0, function () {
            var query, url, json;
            var _a;
            return __generator(this, function (_b) {
                switch (_b.label) {
                    case 0:
                        query = {
                            id: params.ids,
                            _embed: (_a = params === null || params === void 0 ? void 0 : params.meta) === null || _a === void 0 ? void 0 : _a.embed,
                        };
                        url = "".concat(apiUrl, "/").concat(resource, "?").concat(stringify(query));
                        return [4 /*yield*/, httpClient(url, { signal: params === null || params === void 0 ? void 0 : params.signal })];
                    case 1:
                        json = (_b.sent()).json;
                        return [2 /*return*/, { data: json }];
                }
            });
        }); },
        getManyReference: function (resource, params) { return __awaiter(void 0, void 0, void 0, function () {
            var _a, page, perPage, _b, field, order, query, url, _c, headers, json, totalString;
            var _d;
            var _e;
            return __generator(this, function (_f) {
                switch (_f.label) {
                    case 0:
                        _a = params.pagination, page = _a.page, perPage = _a.perPage;
                        _b = params.sort, field = _b.field, order = _b.order;
                        query = __assign(__assign({}, fetchUtils.flattenObject(params.filter)), (_d = {}, _d[params.target] = params.id, _d._sort = field, _d._order = order, _d._start = (page - 1) * perPage, _d._end = page * perPage, _d._embed = (_e = params === null || params === void 0 ? void 0 : params.meta) === null || _e === void 0 ? void 0 : _e.embed, _d));
                        url = "".concat(apiUrl, "/").concat(resource, "?").concat(stringify(query));
                        return [4 /*yield*/, httpClient(url, {
                                signal: params === null || params === void 0 ? void 0 : params.signal,
                            })];
                    case 1:
                        _c = _f.sent(), headers = _c.headers, json = _c.json;
                        if (!headers.has('x-total-count')) {
                            throw new Error('The X-Total-Count header is missing in the HTTP Response. The jsonServer Data Provider expects responses for lists of resources to contain this header with the total number of results to build the pagination. If you are using CORS, did you declare X-Total-Count in the Access-Control-Expose-Headers header?');
                        }
                        totalString = headers.get('x-total-count').split('/').pop();
                        if (totalString == null) {
                            throw new Error('The X-Total-Count header is invalid in the HTTP Response.');
                        }
                        return [2 /*return*/, { data: json, total: parseInt(totalString, 10) }];
                }
            });
        }); },
        update: function (resource, params) { return __awaiter(void 0, void 0, void 0, function () {
            var json;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, httpClient("".concat(apiUrl, "/").concat(resource, "/").concat(params.id), {
                            method: 'PUT',
                            body: JSON.stringify(params.data),
                        })];
                    case 1:
                        json = (_a.sent()).json;
                        return [2 /*return*/, { data: json }];
                }
            });
        }); },
        // json-server doesn't handle filters on UPDATE route, so we fallback to calling UPDATE n times instead
        updateMany: function (resource, params) { return __awaiter(void 0, void 0, void 0, function () {
            var responses;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, Promise.all(params.ids.map(function (id) {
                            return httpClient("".concat(apiUrl, "/").concat(resource, "/").concat(id), {
                                method: 'PUT',
                                body: JSON.stringify(params.data),
                            });
                        }))];
                    case 1:
                        responses = _a.sent();
                        return [2 /*return*/, { data: responses.map(function (_a) {
                                    var json = _a.json;
                                    return json.id;
                                }) }];
                }
            });
        }); },
        create: function (resource, params) { return __awaiter(void 0, void 0, void 0, function () {
            var json;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, httpClient("".concat(apiUrl, "/").concat(resource), {
                            method: 'POST',
                            body: JSON.stringify(params.data),
                        })];
                    case 1:
                        json = (_a.sent()).json;
                        return [2 /*return*/, { data: __assign(__assign({}, params.data), json) }];
                }
            });
        }); },
        delete: function (resource, params) { return __awaiter(void 0, void 0, void 0, function () {
            var json;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, httpClient("".concat(apiUrl, "/").concat(resource, "/").concat(params.id), { method: 'DELETE' })];
                    case 1:
                        json = (_a.sent()).json;
                        return [2 /*return*/, { data: json }];
                }
            });
        }); },
        // json-server doesn't handle filters on DELETE route, so we fallback to calling DELETE n times instead
        deleteMany: function (resource, params) { return __awaiter(void 0, void 0, void 0, function () {
            var responses;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, Promise.all(params.ids.map(function (id) {
                            return httpClient("".concat(apiUrl, "/").concat(resource, "/").concat(id), {
                                method: 'DELETE',
                            });
                        }))];
                    case 1:
                        responses = _a.sent();
                        return [2 /*return*/, { data: responses.map(function (_a) {
                                    var json = _a.json;
                                    return json.id;
                                }) }];
                }
            });
        }); },
    });
});
